File: EntityFX.NetBenchamarks.Core/Generic/RandomMemoryBenchmarkBase.cs
using System;
using System.Linq;
using System.Diagnostics;

namespace EntityFX.NetBenchmark.Core.Generic
{
    public class RandomMemoryBenchmarkResult
    {
        public double Average { get; set; }

        public string Output { get; set; }
    }

    public abstract class RandomMemoryBenchmarkBase<TResult>: BenchmarkBase<TResult>
    {
        private Random random;

        Writer output = new Writer();

        protected bool UseConsole { get => output.UseConsole; set => output.UseConsole = value; }

        public RandomMemoryBenchmarkBase(bool printToConsole = true)
        {
            Iterrations = 500000;
            Ratio = 2;
            random = new Random((int)DateTime.Now.Ticks);
            UseConsole = printToConsole;
        }

        public override void Warmup(double aspect = 0.05)
        {
            UseConsole = false;
            base.Warmup(aspect);
            UseConsole = true;
        }

        public RandomMemoryBenchmarkResult BenchRandomMemory() 
        {
            var int4k = MeasureArrayRandomRead(1000);
            output.WriteLine($"Random int 4k: {int4k.Item1.ToString("F2")} MB/s");
            var int512k = MeasureArrayRandomRead(131072);
            output.WriteLine($"Random int 512k: {int512k.Item1.ToString("F2")} MB/s");
            var int8m = MeasureArrayRandomRead(2097152);
            output.WriteLine($"Random int 8M: {int8m.Item1.ToString("F2")} MB/s");

            var long4k = MeasureArrayRandomLongRead(1000);
            output.WriteLine($"Random long 4k: {long4k.Item1.ToString("F2")} MB/s");
            var long512k = MeasureArrayRandomLongRead(131072);
            output.WriteLine($"Random long 512k: {long512k.Item1.ToString("F2")} MB/s");
            var long8m = MeasureArrayRandomLongRead(2097152);
            output.WriteLine($"Random long 8M: {long8m.Item1.ToString("F2")} MB/s");

            var avg = (new double[] { int4k.Item1, int512k.Item1, int8m.Item1, long4k.Item1, long512k.Item1, long8m.Item1 }).Average();
            output.WriteLine($"Average: {avg.ToString("F2")} MB/s");

            return new RandomMemoryBenchmarkResult() 
            {
                Average = avg,
                Output = output.Output
            };
        }

        protected Tuple<double, int> MeasureArrayRandomRead(int size)
        {
            int I = 0;
            var sw = new Stopwatch();
            var array = Enumerable.Range(0, size).Select( r => random.Next(int.MinValue, int.MaxValue)).ToArray();
            var end = array.Length - 1;
            var indexes = Enumerable.Range(0, end).Select( r => random.Next(0, end)).ToArray();
            var iterInternal = Iterrations / (size / 1000);
            for (int idx = 0; idx < end; idx++)
            {
                I = array[idx];
            }
            sw.Start();
            for (int i = 0; i < iterInternal; i++)
            {
                foreach(var idx in indexes)
                {
                    I = array[idx];
                }
            }
            sw.Stop();
            
            return new Tuple<double, int>((long)iterInternal * array.Length * 4 / sw.Elapsed.TotalSeconds / 1024 / 1024, I);
        }

        protected Tuple<double, long> MeasureArrayRandomLongRead(int size)
        {
            long L = 0;
            var sw = new Stopwatch();
            var array = Enumerable.Range(0, size).Select( r => (long)random.Next(int.MinValue, int.MaxValue)).ToArray();
            var end = array.Length - 1;
            var indexes = Enumerable.Range(0, end).Select( r => random.Next(0, end)).ToArray();
            var iterInternal = Iterrations / (size / 1000);
            for (int idx = 0; idx < end; idx++)
            {
                L = array[idx];
            }
            sw.Start();
            for (int i = 0; i < iterInternal; i++)
            {
                foreach(var idx in indexes)
                {
                    L = array[idx];
                }
            }
            sw.Stop();
            
            return new Tuple<double, long>((long)iterInternal * array.Length * 8 / sw.Elapsed.TotalSeconds / 1024 / 1024, L);
        }
    }
}

File: src/python/entityfx/random_memory_benchmark_base.py

from entityfx.benchmark_base import BenchmarkBase
from entityfx.writer import Writer
import random
import math
import time
from statistics import mean
import sys

class RandomMemoryBenchmarkBase(BenchmarkBase):
    
    
    def __init__(self, writer: Writer, print_to_console : bool=True) -> None:
        super().__init__(writer, print_to_console)

        random.seed()
        
        self.__output = Writer()
        self._iterrations = 500000
        self.ratio = 2
        self._use_console(print_to_console)
    
    def warmup(self, aspect : float=.05) -> None:
        self._use_console(False)
        super().warmup(aspect)
        self._use_console(True)
    
    def benchRandomMemory(self) :
        int4k = self._measureArrayRandomRead(1024)
        self._output.write_line("Random int 4k: {0:5.2f} MB/s", int4k[0])
        int512k = self._measureArrayRandomRead(131072)
        self._output.write_line("Random int 512k: {0:5.2f} MB/s", int512k[0])
        int8m = self._measureArrayRandomRead(2097152)
        self._output.write_line("Random int 8m: {0:5.2f} MB/s", int8m[0])

        long4k = self._measureArrayRandomLongRead(512)
        self._output.write_line("Random long 4k: {0:5.2f} MB/s", long4k[0])
        long512k = self._measureArrayRandomLongRead(65536)
        self._output.write_line("Random long 512k: {0:5.2f} MB/s", long512k[0])
        long8m = self._measureArrayRandomLongRead(1048576)
        self._output.write_line("Random long 8m: {0:5.2f} MB/s", long8m[0])

        avg = mean([int4k[0], int512k[0], int8m[0], long4k[0], long512k[0], long8m[0] ])
        
        return { "Average" : avg, "Output" : self.__output.output }
    
    def _measureArrayRandomRead(self, size) :
        i = 0

        array0_ = list(map(lambda x: random.randint(-2147483647, 2147483647), range(0, size)))
        end = len(array0_) - 1

        indexes = list(map(lambda x: random.randint(0, end), range(0, end)))

        k0 = math.floor(size / 1024)
        k1 = 1 if k0 == 0 else k0

        iter_internal = math.floor(self._iterrations / k1)
        iter_internal = 1 if iter_internal == 0 else iter_internal

        idx = 0
        while idx < end: 
            i = (array0_[idx])
            idx += 1

        start = time.time()
        it = 0


        while it < iter_internal:
            for idx in indexes:
                i = (array0_[idx])
            it += 1


        elapsed = time.time() - start
        return (iter_internal * len(array0_) * 4 / elapsed / 1024 / 1024, i)
    
    def _measureArrayRandomLongRead(self, size) :
        i = 0

        array0_ = list(map(lambda x: random.randint(-2147483647, 2147483647), range(0, size)))
        end = len(array0_) - 1

        indexes = list(map(lambda x: random.randint(0, end), range(0, end)))

        k0 = math.floor(size / 1024)
        k1 = 1 if k0 == 0 else k0

        iter_internal = math.floor(self._iterrations / k1)
        iter_internal = 1 if iter_internal == 0 else iter_internal

        idx = 0
        while idx < end: 
            i = (array0_[idx])
            idx += 1

        start = time.time()
        it = 0


        while it < iter_internal:
            for idx in indexes:
                i = (array0_[idx])
            it += 1


        elapsed = time.time() - start
        return (iter_internal * len(array0_) * 4 / elapsed / 1024 / 1024, i)